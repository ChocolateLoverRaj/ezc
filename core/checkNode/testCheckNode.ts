import EnumItemWithData from '../EnumItemWithData'
import coreNodeFns from '../nodeFns/coreNodeFns'
import TryParseNode from '../parseNode/TryParseNode'
import stringToTokensStream from './stringToTokensStream'

const testCheckNode = async (
  parseNode: TryParseNode<EnumItemWithData>,
  code: string
): Promise<void> => {
  const { success, result } = await parseNode(stringToTokensStream(code))
  if (!success) throw new Error('Error parsing node')
  const { type, data } = result.node
  expect(coreNodeFns.get(type.enum)?.get(type.id)?.check?.({
    allNodeFns: coreNodeFns,
    nodeData: data,
    parentNodes: []
  })).toMatchSnapshot()
}

export default testCheckNode
