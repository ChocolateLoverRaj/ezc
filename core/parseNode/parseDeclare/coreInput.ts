import coreTypeParsers from '../coreTypeParsers'
import coreKeyWordsToInputFlags from '../parseInputFlags/coreKeyWordsToInputFlags'
import Input from './Input'

const coreInput: Input = {
  typeParsers: coreTypeParsers,
  keyWordsToInputFlags: coreKeyWordsToInputFlags
}

export default coreInput
