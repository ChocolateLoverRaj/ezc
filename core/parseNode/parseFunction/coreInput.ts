import coreTypeParsers from '../coreTypeParsers'
import coreInstructionParsers from '../parseBlock/coreInstructionParsers'
import parseBlock from '../parseBlock/parseBlock'
import coreKeyWordsToInputFlags from '../parseInputFlags/coreKeyWordsToInputFlags'
import Input from './Input'

const coreInput: Input = {
  typeParsers: coreTypeParsers,
  keyWordsToInputFlags: coreKeyWordsToInputFlags,
  parseBlock: parseBlock(coreInstructionParsers)
}

export default coreInput
