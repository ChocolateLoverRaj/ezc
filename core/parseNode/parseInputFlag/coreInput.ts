import coreKeyWordsToInputFlags from '../parseInputFlags/coreKeyWordsToInputFlags'
import Input from './Input'

const coreInput: Input = coreKeyWordsToInputFlags

export default coreInput
