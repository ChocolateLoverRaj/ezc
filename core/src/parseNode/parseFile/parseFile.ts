import splitAsyncIterator from 'util/dist/splitAsyncIterator/splitAsyncIterator.js'
import CoreNodesWithData from '../CoreNodesWithData'
import CoreNodeType from '../CoreNodeType'
import TryParseNode from '../TryParseNode'
import skipSplittedAsyncIterator from 'util/dist/splitAsyncIterator/skip.js'
import Input from './Input'
import EnumItemWithData from '../../EnumItemWithData'
import tryNodeParsers from '../tryNodeParsers/tryNodeParsers'

const parseFile = (subNodeParsers: Input): TryParseNode<CoreNodesWithData[CoreNodeType.FILE]> =>
  async stream => {
    const type = {
      enum: CoreNodeType,
      id: CoreNodeType.FILE
    }
    const splittedIterator = splitAsyncIterator(stream[Symbol.asyncIterator]())
    let length = 0
    const skip = (count: number): void => {
      skipSplittedAsyncIterator(splittedIterator, count)
      length += count
    }
    const nodes: EnumItemWithData[] = []
    while (true) {
      const parseNodeResult = await tryNodeParsers(subNodeParsers)(splittedIterator.asyncIterable)
      if (!parseNodeResult.success) {
        // Return undefined if there is a token which can't be parsed
        const { done } = await splittedIterator.asyncIterable[Symbol.asyncIterator]().next()
        if (done === true) break
        else {
          return {
            success: false,
            result: {
              type,
              index: length,
              message: "Couldn't parse any sub-node",
              subAttempts: parseNodeResult.result
            }
          }
        }
      }
      nodes.push(parseNodeResult.result.node)
      skip(parseNodeResult.result.length)
    }

    return {
      success: true,
      result: {
        node: {
          type,
          data: nodes
        },
        length
      }
    }
  }

export default parseFile
