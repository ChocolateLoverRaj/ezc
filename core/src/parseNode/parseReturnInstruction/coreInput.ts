import coreTypeParsers from '../coreTypeParsers'
import coreValueParsers from '../coreValueParsers'
import Input from './Input'

const coreInput: Input = {
  typeParsers: coreTypeParsers,
  valueParsers: coreValueParsers
}

export default coreInput
