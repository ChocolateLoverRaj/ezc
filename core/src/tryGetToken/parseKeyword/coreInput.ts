import EnumItem from '../../EnumItem'
import coreUnparseKeyWordInput from '../../unparseToken/unparseKeyWord/coreInput'
import Input from './Input'

const coreInput: Input = new Map([...coreUnparseKeyWordInput]
  .flatMap(([_enum, map]) => [...map]
    .map<[string, EnumItem]>(([id, string]) => [string, { enum: _enum, id }])))

export default coreInput
