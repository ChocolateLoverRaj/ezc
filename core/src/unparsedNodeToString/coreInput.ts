import coreUnparseTokenInput from '../unparseToken/unparseToken/coreInput'
import coreNodeFns from '../nodeFns/coreNodeFns'
import Input from './Input'

const coreInput: Input = {
  unparseTokenInput: coreUnparseTokenInput,
  allNodeFns: coreNodeFns
}

export default coreInput
