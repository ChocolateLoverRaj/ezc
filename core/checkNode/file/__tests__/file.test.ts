import coreInput from '../../../parseNode/parseFile/coreInput'
import parseFile from '../../../parseNode/parseFile/parseFile'
import testCheckNode from '../../testCheckNode'

test('1 duplicate identifier with 2 duplicates', async () => {
  await testCheckNode(parseFile(coreInput),
`declare void @someFn ()
declare void @someFn ()`)
})

test('1 duplicate identifier with 3 duplicates', async () => {
  await testCheckNode(parseFile(coreInput),
`declare void @doStuff ()
declare void @doStuff ()
declare void @doStuff ()`)
})

test('2 duplicate identifiers with 2 duplicates', async () => {
  await testCheckNode(parseFile(coreInput),
`declare void @a ()
declare void @b ()
declare void @a ()
declare void @b ()`)
})
