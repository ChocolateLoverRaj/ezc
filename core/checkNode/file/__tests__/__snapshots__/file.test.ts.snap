// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 duplicate identifier with 2 duplicates 1`] = `
Array [
  Object {
    "locations": Array [
      Array [
        0,
        1,
      ],
      Array [
        1,
        1,
      ],
    ],
    "message": "Duplicate identifier: @someFn",
  },
]
`;

exports[`1 duplicate identifier with 3 duplicates 1`] = `
Array [
  Object {
    "locations": Array [
      Array [
        0,
        1,
      ],
      Array [
        1,
        1,
      ],
      Array [
        2,
        1,
      ],
    ],
    "message": "Duplicate identifier: @doStuff",
  },
]
`;

exports[`2 duplicate identifiers with 2 duplicates 1`] = `
Array [
  Object {
    "locations": Array [
      Array [
        0,
        1,
      ],
      Array [
        2,
        1,
      ],
    ],
    "message": "Duplicate identifier: @a",
  },
  Object {
    "locations": Array [
      Array [
        1,
        1,
      ],
      Array [
        3,
        1,
      ],
    ],
    "message": "Duplicate identifier: @b",
  },
]
`;
